src/relations/mysql_tls.py

```diff
@@ -0,0 +1,237 @@
+        if param is None:
+            key = generate_private_key()
+        else:
+            key = self._parse_tls_file(param)
+
+        csr = generate_csr(
+            private_key=key,
+            subject=self.charm.unit_peer_data["instance-hostname"].split(":")[0],
+            organization=self.charm.app.name,
+            sans=self._get_sans(),
+        )
+
+        # store secrets
+        self.charm.set_secret(SCOPE, "key", key.decode("utf-8"))
+        self.charm.set_secret(SCOPE, "csr", csr.decode("utf-8"))
+
+        if self.charm.model.get_relation(TLS_RELATION):
+            self.certs.request_certificate_creation(certificate_signing_request=csr)
+
+    @staticmethod
+    def _parse_tls_file(raw_content: str) -> bytes:
+        """Parse TLS files from both plain text or base64 format."""
+        if re.match(r"(-+(BEGIN|END) [A-Z ]+-+)", raw_content):
+            return re.sub(
+                r"(-+(BEGIN|END) [A-Z ]+-+)",
+                "\n\\1\n",
+                raw_content,
+            ).encode("utf-8")
+        return base64.b64decode(raw_content)
+
+    def _get_sans(self) -> List[str]:
+        """Create a list of DNS names for a unit.
+
+        Returns:
+            A list representing the hostnames of the unit.
+        """
+        unit_id = self.charm.unit.name.split("/")[1]
+        return [
+            f"{self.charm.app.name}-{unit_id}",
+            socket.getfqdn(),
+            str(self.charm.model.get_binding(self.charm.peers).network.bind_address),
+        ]
+
+    def get_tls_content(self) -> Tuple[Optional[str], Optional[str], Optional[str]]:
+        """Retrieve TLS content.
+
+        Return TLS files as required by mysql.
+
+        Returns:
+            A tuple of strings with the content of server-key, ca and server-cert
+        """
+        ca = self.charm.get_secret(SCOPE, "ca")
+        chain = self.charm.get_secret(SCOPE, "chain")
+        ca_file = chain or ca
+
+        key = self.charm.get_secret(SCOPE, "key")
+        cert = self.charm.get_secret(SCOPE, "cert")
+        return key, ca_file, cert
+
+    def push_tls_files_to_workload(self) -> None:
+        """Push TLS files to unit."""
+        ssl_key, ssl_ca, ssl_cert = self.get_tls_content()
+
+        if ssl_key:
+            write_content_to_file(
+                f"{MYSQL_DATA_DIR}/{TLS_SSL_KEY_FILE}", ssl_key, permission=0o400
+            )
+
+        if ssl_ca:
+            write_content_to_file(f"{MYSQL_DATA_DIR}/{TLS_SSL_CA_FILE}", ssl_ca, permission=0o400)
+
+        if ssl_cert:
+            write_content_to_file(
+                f"{MYSQL_DATA_DIR}/{TLS_SSL_CERT_FILE}", ssl_cert, permission=0o400
+            )
+
+    @staticmethod
+    def create_tls_config_file() -> None:
+        """Render TLS template directly to file.
+
+        Render and write TLS enabling config file from template.
+        """
+        with open("templates/tls.cnf", "r") as template_file:
+            template = Template(template_file.read())
```

Comment:
should we use Jinja instead? (that is what we use for templates in other projects and elsewhere)

Reply:
I prefer `Template` over jinja because is lightweight and a builtin, more appropriate for simple tag replacement. Not passionate about it tho. Tbh i forgot to check what we were using, but I propose to change elsewhere or here to use a single solution.

Reply:
on this charm, there's no jinja dependency - so I would prefer to keep as is